samcli/lib/build/app_builder.py

```diff
@@ -125,9 +126,19 @@ def build(self):
         dict
             Returns the path to where each resource was built as a map of resource's LogicalId to the path string
         """
+
+        async_context = None
+        if self._parallel:
+            async_context = AsyncContext()
+
         build_graph = self._get_build_graph()
-        result = self._build_functions(build_graph)
-        result.update(self._build_layers(build_graph))
+        result = self._build_functions(async_context, build_graph)
```

Comment:
The way this code flow works is pretty confusing - if I'm understanding it right, this returns an empty hash when you're doing async, and in that case it updates the async_context as a side effect. It really feels like we should have a true fork in the logic with different functions in this case, this is going to be difficult to understand/test/maintain as-is in my view. At minimum we should be clearly commenting and documenting this change in approach.